src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/W3CPropagator.cs

```diff
@@ -0,0 +1,741 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace System.Diagnostics
+{
+    internal sealed class W3CPropagator : DistributedContextPropagator
+    {
+        internal static DistributedContextPropagator Instance { get; } = new W3CPropagator();
+
+        private const int MaxBaggageEntriesToEmit = 64;     // Suggested by W3C specs
+        private const int MaxBaggageEncodedLength = 8192;   // Suggested by W3C specs
+        private const int MaxTraceStateEncodedLength = 256; // Suggested by W3C specs
+
+        private const char Equal = '=';
+        private const char Percent = '%';
+        private const char Replacement = '\uFFFD'; // �
+
+        public override IReadOnlyCollection<string> Fields { get; } = new ReadOnlyCollection<string>(new[] { TraceParent, TraceState, Baggage });
+
+        public override void Inject(Activity? activity, object? carrier, PropagatorSetterCallback? setter)
+        {
+            if (activity is null || setter is null || activity.IdFormat != ActivityIdFormat.W3C)
+            {
+                return;
+            }
+
+            string? id = activity.Id;
+            if (id is null)
+            {
+                return;
+            }
+
+            setter(carrier, TraceParent, id);
+            if (!string.IsNullOrEmpty(activity.TraceStateString))
+            {
+                InjectTraceState(activity.TraceStateString, carrier, setter);
+            }
+
+            InjectW3CBaggage(carrier, activity.Baggage, setter);
+        }
+
+        public override void ExtractTraceIdAndState(object? carrier, PropagatorGetterCallback? getter, out string? traceId, out string? traceState)
+        {
+            if (getter is null)
+            {
+                traceId = null;
+                traceState = null;
+                return;
+            }
+
+            getter(carrier, TraceParent, out traceId, out _);
+            if (IsInvalidTraceParent(traceId))
+            {
+                traceId = null;
+            }
+
+            getter(carrier, TraceState, out string? traceStateValue, out _);
+            traceState = ValidateTraceState(traceStateValue);
+        }
+
+        public override IEnumerable<KeyValuePair<string, string?>>? ExtractBaggage(object? carrier, PropagatorGetterCallback? getter)
```

Comment:
More a design question than implementation:

Rejecting the correlation-context header as input may make adoption of the new version harder because it requires all processes to be switched at the same time. We could allow the correlation-context header or baggage header as input to increase compatibility while still emitting only the baggage header. Thoughts?

Reply:
I was debating myself on that and decided to go with a more restricted model and we'll have the chance to support compatibility if we see a demand for it. If you think we should start with the relaxed model, I'll be ok too. If we go with the suggestion, the baggage entries will still be read according to W3C specs though. Let me know what do you think.

Reply:
I'd suggest accepting correlation-context in cases where it is present and baggage is not present. I'd worry that even if we don't get complaints about compat during the preview builds we still might get complaints after we ship RTM and people really start to deploy .NET 10 into more complex production systems.

What do you think @samsp-msft?

Reply:
Thanks for the feedback. I have supported `correlation-context` when extracting the baggage.